File: HashTable/HashTable.py
# This binarySearchTree structure is written by Moh.Massoud in 06July18
class node:
    def __init__(self, key, value, d=False):
        self.key = key
        self.value = value
        self.deleted = d

    def __str__(self):
        return "{} : {}".format(self.key, self.value)


class HashTable:
    def __init__(self):
        self.capacity = 50
        self.size = 0
        self.list = [None] * self.capacity

    def _genStr(self):
        str = ""
        for i in self.list:
            if (i != None):
                str += "{} : {}, ".format(i.key, i.value)
        return str[:-2]

    def __str__(self):
        return "[" + self._genStr() + "]"
               # + "\nCapacity of HashTable is {}".format(self.capacity)

    def _hash(self, key):
        sum = 0
        for i in key:
            sum += ord(i)
        return (((sum * 353) % 1000000007) * 23) % self.capacity

    def insert(self, key, value, notOperation=True):
        index = self._hash(key)
        while (self.list[index] != None and self.list[index].deleted != True):
            index += 1
        self.list[index] = node(key, value)
        if (notOperation):
            self.size += 1
        if (self.size >= self.capacity / 2 or index > self.capacity):
            self._doubleCapacity()
        elif (self.size * 4 < self.capacity):
            self._halfenCapacity()

    def _doubleCapacity(self):
        listCopy = self.list.copy()
        self.capacity = self.capacity * 2
        self.list = [None] * self.capacity
        for x in listCopy:
            if x != None:
                self.insert(x.key, x.value, False)

    def _halfenCapacity(self):
        listCopy = self.list.copy()
        self.capacity = self.capacity // 2
        self.list = [None] * self.capacity
        for x in listCopy:
            if x != None:
                self.insert(x.key, x.value, False)

    def find(self, key):
        idx = self._find(key)
        if (idx != -1):
            return self.list[idx]
        else:
            print("Node isn't found!")
            return -1

    def _find(self, key):
        index = self._hash(key)
        while (self.list[index] != None and self.list[index].key != key):
            index += 1
        if (self.list[index] != None and self.list[index].key == key):
            return index
        return -1

    def remove(self, key):
        idx = self._find(key)
        if (idx != -1):
            self.list[idx] = None
            self.size -= 1
            if (self.size * 4 < self.capacity):
                self._halfenCapacity()
        else:
            print("key is not found!!")

    def getValue(self, key):
        idx = self._find(key)
        if (idx != -1):
            return self.list[idx].value
        else:
            print("key is not found!!")


h = HashTable()
h.insert("map1", 1)
print(h)
h.insert("map2", 2)
print(h)
h.insert("map3", 3)
print(h)
h.insert("map4", 4)
print(h)
h.remove("map1")
print(h)
h.insert("map1", 1)
print(h)
h.insert("map5", 5)
print(h)
h.insert("map6", 6)
print(h.getValue("map5"))
print(h)
h.remove("map7")
print(h)
h.remove("map5")
print(h)
h.remove("map2")
print(h)
h.remove("map1")
print(h)
h.remove("map6")
print(h)
h.remove("map3")
print(h)
h.remove("map4")
print(h)
print(h.getValue("map8"))

File: HashTable/HashTable.java
public class HashTable<M,N>{
    class Node<M,N>{

        private M key;
        private N value;
        private boolean deleted = false;

        public Node(M key,N value){
            this.key = key;
            this.value = value;
        }

        public void delete(){
            this.deleted = true;
        }

        public boolean isDeleted(){
            return deleted;
        }

        public M getKey() {
            return key;
        }

        public N getValue() {
            return value;
        }

        public void setValue(N value) {
            this.value = value;
        }
    }

    private int capacity = 4;
    private int size;
    private Node table[];

    public HashTable(){
        table = new Node[capacity];
    }

    private int hash(M key){
        return key.hashCode() % capacity;
    }

    public void insert(M key, N value){
        int found = find(key);

        if(found != -1){
            table[found].setValue(value);
            return;
        }

        int hashedKey = hash(key), index = hashedKey, i = 0;
        while(table[index] != null && !table[index].isDeleted()){
            index = (hashedKey + i) % capacity;
            i++;
        }
        table[index] = new Node(key, value);
        size++;
        if(size > capacity/2)
            doubleCapacity();
        else if (size * 4 < capacity)
            halfenCapacity();
    }

    private void halfenCapacity() {
        this.capacity /= 2;
        Node oldTable[] = table.clone();
        this.table = new Node[capacity];
        for (Node e : oldTable) {
            if (e != null && !e.isDeleted())
                insert((M)e.getKey(), (N)e.getValue());
        }
    }

    private void doubleCapacity() {
        this.capacity *= 2;
        Node oldTable[] = table.clone();
        this.table = new Node[capacity];
        for (Node e : oldTable) {
            if (e != null && !e.isDeleted())
                insert((M)e.getKey(), (N)e.getValue());
        }
    }

    private int find(M key){
        int hashedKey = hash(key), index = hashedKey, i = 0;
        while(table[index] != null && table[index].getKey() != key){
            index = (hashedKey + i) % capacity;
            i++;
        }
        if(table[index] != null && !table[index].isDeleted() && table[index].getKey() == key)
            return index;
        return -1;
    }

    public N getValue(M key){
        int index = find(key);
        if(index != -1)
            return (N)table[index].getValue();
        return null;
    }

    public N remove(M key){
        int index = find(key);
        if(index != -1){
            N value = (N)table[index].getValue();
            table[index].delete();
            size--;
            if (size * 4 < capacity)
                halfenCapacity();
            return value;
        }
        System.out.println("key is not found!!");
        return null;
    }

    private String genStr(){
        StringBuilder stringBuilder = new StringBuilder();
        for(Node e:table){
            if(e != null && !e.isDeleted()){
                stringBuilder.append(String.format("%s : %s, ", e.getKey(), e.getValue()));
            }
        }
        if (stringBuilder.length() > 1) {
            stringBuilder.setLength(stringBuilder.length() - 2);
        }
        return stringBuilder.toString();
    }

    @Override
    public String toString() {
        return "[" + genStr() + "]";
    }

    public static void main(String args[]){
        HashTable<String, Integer> hashTable = new HashTable<>();

        hashTable.insert("map1", 1);
        System.out.println(hashTable);
        hashTable.insert("map2", 2);
        System.out.println(hashTable);
        hashTable.insert("map3", 3);
        System.out.println(hashTable);
        hashTable.insert("map4", 4);
        System.out.println(hashTable);

        System.out.println(hashTable.remove("map1"));
        System.out.println(hashTable);

        hashTable.insert("map1",1);
        System.out.println(hashTable);
        hashTable.insert("map5",5);
        System.out.println(hashTable);
        hashTable.insert("map6",6);
        System.out.println(hashTable.getValue("map5"));
        System.out.println(hashTable);
        hashTable.remove("map7");
        System.out.println(hashTable);
        hashTable.remove("map5");
        System.out.println(hashTable);
        hashTable.remove("map2");
        System.out.println(hashTable);
        hashTable.remove("map1");
        System.out.println(hashTable);
        hashTable.remove("map6");
        System.out.println(hashTable);
        hashTable.remove("map3");
        System.out.println(hashTable);
        hashTable.remove("map4");
        System.out.println(hashTable);
        System.out.println(hashTable.getValue("map8"));
        hashTable.insert("map1",12);
        System.out.println(hashTable);
        hashTable.insert("map1",16);
        System.out.println(hashTable);
        hashTable.remove("map1");
        System.out.println(hashTable);
        hashTable.insert("map1",13);
        System.out.println(hashTable);
    }
}
